### PR TITLE
[compiler-rt] Implement __extendxftf2 and __trunctfxf2 for x86_64

### PR DESCRIPTION
**Pull request by alexander-shaposhnikov**
This patch implements __extendxftf2 (long double -> f128) and
__trunctfxf2 (f128 -> long double) on x86_64.
This is a preparation to unblock https://reviews.llvm.org/D53608,
We intentionally do not modify compiler-rt/lib/builtins/fp_lib.h in this PR
(in particular, to limit the scope and avoid exposing other functions on X86_64 in this PR).
Instead, TODOs were added to use fp_lib.h once it is available.

Test plan:
1. ninja check-compiler-rt (verified on X86_64 and on Aarch64)
In particular, new tests (extendxftf2_test.c and trunctfxf2_test.c) were added.
2. compared the results of conversions with what other compilers (gcc) produce.

### CODE REVIEW

**Comment by alexander-shaposhnikov** [comment]
@efriedma-quic

**Comment by echristo** [comment]
Adding @erichkeane here as well to help review.

**Comment by erichkeane** [comment]
> Adding @erichkeane here as well to help review.

I have no idea about this, I actually have no idea what these do.  Were you perhaps thinking of someone else?

**Review by echristo — APPROVED** [review]
This looks reasonable to me. I've added Erich in case he has any comments before merging.

**Comment by efriedma-quic** [comment]
I don't see any tests for denormal numbers?

I'm a little concerned that mixing together fp80 and non-fp80 codepaths is going to be confusing for anyone unfamiliar with fp80 because the rules for fp80 significands are different... can you try to pull out the differences into helper methods, so it's more clear what's the IEEE codepath, and what's the non-IEEE codepath?

**Comment by echristo** [comment]
Sorry, was thinking from the x86 perspective. If you don't feel comfortable
that's ok, Eli has a lot of background here. Just wanted to make sure you
were aware. :)

File: compiler-rt/lib/builtins/fp_extend_impl.inc
Comment (alexander-shaposhnikov) [inline compiler-rt/lib/builtins/fp_extend_impl.inc:105]: for extension this is the only place where F80->F128 is special-cased,
since it uses various things declared above - decided not to factor it out, but added a comment instead.

**Comment by alexander-shaposhnikov** [comment]
1. Added more tests (for denormal numbers).
2. Refactored/Unified the codepaths for f_80 vs usual f_N.

**Comment by MaskRay** [comment]
I will be out of town for about 2 weeks but happy to take a look after the trip...

File: compiler-rt/lib/builtins/fp_extend.h
Comment (efriedma-quic) [inline compiler-rt/lib/builtins/fp_extend.h:54]: To be specific, I think the approach here is making the patch much harder to understand than is really necessary. 
 I think defining two different constants representing the size of the significand, and integrating the logic to handle these directly into the routines, is confusing.

I'd like to see a helper function "extract significant fraction bits" or something like that.  And a function to construct a float given the exponent, and significant fraction bits.  Then the rest of the code can use those helpers, instead of making it directly aware of the extra bit in fp80 representations.

File: compiler-rt/lib/builtins/fp_extend_impl.inc
Comment (efriedma-quic) [reply compiler-rt/lib/builtins/fp_extend_impl.inc:105]: If I understand correctly, the key points here are:

- The exponent fields are identical
- This is a denormal number, so all the non-significand bits are zero.

So we just need to shift the bits.  (And this code could be used with other floating-point formats, e.g. converting bfloat16->float.  The differences between fp80 and other formats aren't relevant.)

Please update the comment to indicate f80->f128 is just a special-case of this general issue.

File: compiler-rt/lib/builtins/fp_extend_impl.inc
Comment (efriedma-quic) [inline compiler-rt/lib/builtins/fp_extend_impl.inc:118]: Can you just use `#else` here?  (Also, explicitly note that this codepath is currently unused for fp80.)

File: compiler-rt/lib/builtins/fp_extend_impl.inc
Comment (alexander-shaposhnikov) [reply compiler-rt/lib/builtins/fp_extend_impl.inc:105]: Correct, 
sure, will update the comment.

File: compiler-rt/lib/builtins/fp_extend_impl.inc
Comment (alexander-shaposhnikov) [reply compiler-rt/lib/builtins/fp_extend_impl.inc:118]: oh, right, will do.

File: compiler-rt/lib/builtins/fp_extend.h
Comment (alexander-shaposhnikov) [reply compiler-rt/lib/builtins/fp_extend.h:54]: ok, I initially started with something similar to what you are suggesting,
but this turned out to be quite far from what the existing template.
However, I agree, this probably would be cleaner, I'll try to udpate the PR ~soon.

Comment (alexander-shaposhnikov) [reply compiler-rt/lib/builtins/fp_extend.h:54]: @efriedma-quic - I'm working on giving it another try, but want to point out - that this would be essentially a rewrite of the existing template (so the changes will be quite invasive). Unfortunately, the current implementation of __extendXfYf2__ (even without any of my changes) is not that simple and doesn't use the model that you'd like to see.

File: compiler-rt/lib/builtins/fp_extend.h
Comment (arichardson) [inline compiler-rt/lib/builtins/fp_extend.h:118]: Does this need to use `__inline` rather than `inline`?

File: compiler-rt/lib/builtins/fp_extend.h
Comment (arichardson) [inline compiler-rt/lib/builtins/fp_extend.h:24]: These constants are already defined inside fp_lib.h. Wouldn't it make more sense to refactor fp_lib.h so that instead of 
```
#if defined SINGLE_PRECISION
...
#define significandBits 23
#elif defined DOUBLE_PRECISION
...
#define significandBits 52
```
we have something like
```
#define sf_significand_bits 23
#define df_significand_bits 52
#define xf_significand_bits 63
#if defined SINGLE_PRECISION
#define significandBits sf_significand_bits
#elif defined DOUBLE_PRECISION
#define significandBits df_significand_bits
```
and then reuse those constants here?

Comment (alexander-shaposhnikov) [reply compiler-rt/lib/builtins/fp_extend.h:24]: It definitely would, but I wanted to apply these changes in multiple phases 
(see also the description)  (partially to limit the scope of this PR and partially to avoid concurrently editing fp_lib.h (so that multiple PRs could make progress/land independently / a bit easier).

File: compiler-rt/lib/builtins/fp_extend.h
Comment (efriedma-quic) [reply compiler-rt/lib/builtins/fp_extend.h:55]: I don't see an issue with making relatively extensive changes here; the old code is pretty fragile and hard to read anyway.  (I mean, we say "invasive", but it's not affecting anything outside the implementation itself.)

Updated extend looks good; it's shorter and easier to read, even with the new functionality.

File: compiler-rt/lib/builtins/fp_extend.h
Comment (efriedma-quic) [reply compiler-rt/lib/builtins/fp_extend.h:118]: We're already using the usual "inline" keyword in other places, so it should be fine to use here.  (I assume the use of __inline is an artifact from when compiler-rt was built in "c89" mode.)

File: compiler-rt/lib/builtins/fp_extend_impl.inc
Comment (arichardson) [inline compiler-rt/lib/builtins/fp_extend_impl.inc:40]: Is this include actually required or left over from debugging?

Comment (alexander-shaposhnikov) [reply compiler-rt/lib/builtins/fp_extend_impl.inc:40]: good catch, no, it's a left over

File: compiler-rt/lib/builtins/fp_extend.h
Comment (alexander-shaposhnikov) [reply compiler-rt/lib/builtins/fp_extend.h:56]: Updated "truncate" as well.

**Comment by arichardson** [comment]
Some of this can be simplified if https://github.com/llvm/llvm-project/pull/68132 lands first, but I guess the cleanups could also be done afterwards?

**Comment by alexander-shaposhnikov** [comment]
@arichardson - yeah, there are opportunities for clean ups, adding these conversions will unblock a few things => so wanted to do them afterwards.

**Comment by alexander-shaposhnikov** [comment]
@efriedma-quic - both "extend" and "truncate" have been updated/refactored.

**Review by arichardson — APPROVED** [review]
No objections from my side, cleanups can be done later.

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Pull request merged**
